File: js/src/string/levenshtein.ts
import { EvaluatorResult } from "../types.js";
import { _runEvaluator } from "../utils.js";

function scorer(outputs: unknown, referenceOutputs: unknown): number {
  if (outputs === null || referenceOutputs === null) {
    throw new Error(
      "Levenshtein distance requires both outputs and reference_outputs"
    );
  }

  // Convert to strings if needed
  const outputStr =
    typeof outputs === "string" ? outputs : JSON.stringify(outputs);
  const referenceStr =
    typeof referenceOutputs === "string"
      ? referenceOutputs
      : JSON.stringify(referenceOutputs);

  // Create a matrix of size (m+1)x(n+1) where m and n are the string lengths
  const m = outputStr.length;
  const n = referenceStr.length;
  const dp: number[][] = Array(m + 1)
    .fill(null)
    .map(() => Array(n + 1).fill(0));

  // Initialize first row and column
  for (let i = 0; i <= m; i++) {
    dp[i][0] = i;
  }
  for (let j = 0; j <= n; j++) {
    dp[0][j] = j;
  }

  // Fill the matrix
  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      if (outputStr[i - 1] === referenceStr[j - 1]) {
        dp[i][j] = dp[i - 1][j - 1];
      } else {
        dp[i][j] = Math.min(
          dp[i - 1][j] + 1, // deletion
          dp[i][j - 1] + 1, // insertion
          dp[i - 1][j - 1] + 1 // substitution
        );
      }
    }
  }

  // Calculate the distance and normalize it to a score between 0 and 1
  const distance = dp[m][n];
  const maxLength = Math.max(m, n);
  return maxLength > 0 ? 1.0 - distance / maxLength : 1.0;
}

/**
 * Evaluates the actual output and reference output for similarity by Levenshtein distance.
 * @param options - The options object containing outputs and reference outputs
 * @returns EvaluatorResult containing match result with score between 0.0 and 1.0,
 * where 1.0 indicates an exact match and lower values indicate greater differences
 */
export async function levenshteinDistance(params: {
  outputs: unknown;
  referenceOutputs?: unknown;
}): Promise<EvaluatorResult> {
  const { outputs, referenceOutputs } = params;

  const getScore = () => scorer(outputs, referenceOutputs);

  return _runEvaluator(
    "levenshtein_distance",
    getScore,
    "levenshtein_distance"
  );
}

File: python/openevals/string/levenshtein.py
import json
from openevals.types import EvaluatorResult
from openevals.utils import _run_evaluator, _arun_evaluator
from typing import Any


def _scorer(outputs: Any, reference_outputs: Any) -> float:
    if outputs is None or reference_outputs is None:
        raise ValueError(
            "Levenshtein distance requires both outputs and reference_outputs"
        )
    if not isinstance(outputs, str):
        outputs = json.dumps(outputs)
    if not isinstance(reference_outputs, str):
        reference_outputs = json.dumps(reference_outputs)
    # Use the shorter string for the row width to use O(min(m, n)) memory.
    m, n = len(outputs), len(reference_outputs)
    if n > m:
        outputs, reference_outputs = reference_outputs, outputs
        m, n = n, m
    previous = list(range(n + 1))
    current = [0] * (n + 1)

    # Each cell only needs the current row and the previous row.
    for i in range(1, m + 1):
        current[0] = i
        for j in range(1, n + 1):
            if outputs[i - 1] == reference_outputs[j - 1]:
                current[j] = previous[j - 1]
            else:
                current[j] = min(
                    previous[j] + 1,  # deletion
                    current[j - 1] + 1,  # insertion
                    previous[j - 1] + 1,  # substitution
                )
        previous, current = current, previous

    # Calculate the distance and normalize it to a score between 0 and 1
    distance = previous[n]
    max_length = max(m, n)
    score = 1.0 - (distance / max_length) if max_length > 0 else 1.0
    return score


def levenshtein_distance(
    *, outputs: Any, reference_outputs: Any, **kwargs: Any
) -> EvaluatorResult:
    """
    Evaluates the actual output and reference output for similarity by Levenshtein distance.

    Args:
        outputs (Any): Outputs to compare
        reference_outputs (Any): Reference outputs to compare

    Returns:
        EvaluatorResult: Contains match result with score between 0.0 and 1.0, where 1.0 indicates
        an exact match and lower values indicate greater differences
    """

    def get_score():
        return _scorer(outputs, reference_outputs)

    res = _run_evaluator(
        run_name="levenshtein_distance",
        scorer=get_score,
        feedback_key="levenshtein_distance",
    )
    if isinstance(res, list):
        return res[0]
    return res


async def levenshtein_distance_async(
    *, outputs: Any, reference_outputs: Any, **kwargs: Any
) -> EvaluatorResult:
    """
    Evaluates the actual output and reference output for similarity by Levenshtein distance.

    Args:
        outputs (Any): Outputs to compare
        reference_outputs (Any): Reference outputs to compare

    Returns:
        EvaluatorResult: Contains match result with score between 0.0 and 1.0, where 1.0 indicates
        an exact match and lower values indicate greater differences
    """

    async def get_score():
        return _scorer(outputs, reference_outputs)

    res = await _arun_evaluator(
        run_name="levenshtein_distance",
        scorer=get_score,
        feedback_key="levenshtein_distance",
    )
    if isinstance(res, list):
        return res[0]
    return res
